src/toxcore/dht/request_queue.rs

```diff
@@ -1,71 +1,86 @@
+impl<T> RequestQueue<T> {
     /// Create new `RequestQueue`.
-    pub fn new(timeout: Duration) -> RequestQueue {
+    pub fn new(timeout: Duration) -> Self {
         RequestQueue {
             ping_map: HashMap::new(),
             timeout,
         }
     }
 
     /// Generate unique non zero request ID.
-    fn generate_ping_id(&self, pk: PublicKey) -> u64 {
+    fn generate_ping_id(&self) -> u64 {
         loop {
-            let ping_id = random_u64();
-            if ping_id != 0 && !self.ping_map.contains_key(&(pk, ping_id)) {
+            let ping_id = gen_ping_id();
+            if !self.ping_map.contains_key(&ping_id) {
                 return ping_id;
             }
         }
     }
 
     /// Generate and store unique non zero request ID. Later this request ID can
     /// be verified with `check_ping_id` function.
-    pub fn new_ping_id(&mut self, pk: PublicKey) -> u64 {
-        let ping_id = self.generate_ping_id(pk);
-        self.ping_map.insert((pk, ping_id), clock_now());
+    pub fn new_ping_id(&mut self, data: T) -> u64 {
+        let ping_id = self.generate_ping_id();
+        self.ping_map.insert(ping_id, (clock_now(), data));
         ping_id
     }
 
-    /// Check whether request ID is correct and not timed out. This function
-    /// removes received request ID so that it can be verified only once.
-    pub fn check_ping_id(&mut self, pk: PublicKey, ping_id: u64) -> bool {
+    /// Check whether request ID is correct and not timed out. When data
+    /// satisfies passed condition this function removes received request ID and
+    /// returns stored data. So a request ID can be verified only once.
+    pub fn check_ping_id<F: FnOnce(&T) -> bool>(&mut self, ping_id: u64, cond: F) -> Option<T> {
         if ping_id == 0 {
-            return false
+            return None;
         }
 
-        match self.ping_map.remove(&(pk, ping_id)) {
-            Some(time) if clock_elapsed(time) <= self.timeout => true,
-            _ => false,
+        if let Entry::Occupied(entry) = self.ping_map.entry(ping_id) {
+            let &(time, ref data) = entry.get();
+            if clock_elapsed(time) <= self.timeout && cond(data) {
+                let (_ping_id, (_time, data)) = entry.remove_entry();
+                Some(data)
+            } else {
+                None
+            }
+        } else {
+            None
         }
     }
 
     /// Remove timed out request IDs.
     pub fn clear_timed_out(&mut self) {
         let timeout = self.timeout;
-        self.ping_map.retain(|&_, &mut time|
+        self.ping_map.retain(|&_, &mut (time, _)|
             clock_elapsed(time) <= timeout
         );
     }
+
+    /// Get not timed out requests stored in this `RequestQueue`.
+    pub fn get_values(&self) -> impl Iterator<Item = (Instant, &T)> {
+        self.ping_map
+            .values()
+            .filter(move |(time, _)| clock_elapsed(*time) <= self.timeout)
+            .map(|&(time, ref data)| (time, data))
```

Comment:
.map(|(time, data)| (*time, data))

Reply:
Why?

Reply:
Because I think it is easier to read than your version.